src/catalua.cpp

```diff
@@ -824,9 +824,90 @@ int call_lua(std::string tocall)
     return err;
 }
 
-void lua_callback(const char *callback_name)
+void lua_delete_global( const char* name ) {
+    lua_State *L = lua_state;
+    lua_pushnil( L );
+    lua_setglobal( L, name );
+}
+
+void lua_callback_cleanup()
+{
+    lua_delete_global( "callback_last" );
+    lua_delete_global( "callback_arg_count" );
+    lua_delete_global( "mapgen_generator_type" );
+    lua_delete_global( "mapgen_terrain_type_id" );
+    lua_delete_global( "mapgen_terrain_coordinates" );
+    lua_delete_global( "skill_increased_source" );
+    lua_delete_global( "skill_increased_id" );
+    lua_delete_global( "skill_increased_level" );
+    lua_delete_global( "mutation_gained" );
+    lua_delete_global( "mutation_lost" );
+    lua_delete_global( "stat_changed" );
+    lua_delete_global( "stat_value" );
+    lua_delete_global( "item_last_worn" );
+    lua_delete_global( "item_last_taken_off" );
+    lua_delete_global( "effect_changed" );
+    lua_delete_global( "effect_intensity" );
+    lua_delete_global( "effect_bodypart" );
+    lua_delete_global( "mission_finished" );
+    lua_delete_global( "mission_assigned" );
+    lua_delete_global( "weather_new" );
+    lua_delete_global( "weather_old" );
+}
+
+void lua_callback_savelast( const char *callback_name )
+{
+    if( lua_state == nullptr ) {
+        return;
+    }
+    lua_State *L = lua_state;
+
+    lua_pushstring( L, callback_name );
+    lua_setglobal( L, "callback_last" );
+}
+
+void lua_callback( const char *callback_name, CallbackArgumentContainer callback_args )
+{
+    if( lua_state == nullptr ) {
+        return;
+    }
+    lua_State *L = lua_state;
+
+    lua_callback_cleanup();
+
+    int callback_arg_count = callback_args.size();
+    lua_pushinteger( L, callback_arg_count );
+    lua_setglobal( L, "callback_arg_count" );
+
+    for( CallbackArgument callback_arg : callback_args ) {
+        std::string callback_arg_type = callback_arg.GetType();
+        if( callback_arg_type == "string" ) {
```

Comment:
It would be much easier to move this whole logic into a member function of `CallbackArgument`.

This would allow direct access to the private members of that class and you would not need all those getter functions.

Reply:
Thanks, I considered this too, but had some troubles with lua headers, when I moved this code to Save() method in CallbackArgument class.